### DISCUSSION TITLE
How to unpack contents of  __root__ automatically when correctly inferring correct object instance when using discriminated union

### DISCUSSION THREAD

**Discussion by RobertasPetrauskas1** [fastapi/fastapi#8606] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum
from typing import Literal, Union

from fastapi import FastAPI
from pydantic import BaseModel, validator, Field

app = FastAPI()


class ItemType(str, Enum):
    BANANA = "banana"
    SOFA = "sofa"
    DEODORANT = "deodorant"


class Item(BaseModel):
    type: ItemType
    name: str
    is_edible: bool


class Banana(Item):
    type: Literal[ItemType.BANANA]
    is_edible = True

    @validator("name")
    def name_validator(cls, v):
        print("Banana name validator called")
        return v


class Sofa(Item):
    type: Literal[ItemType.SOFA]
    is_edible = False

    @validator("name")
    def name_validator(cls, v):
        print("Sofa name validator called")
        return v


class Deodorant(Item):
    type: Literal[ItemType.DEODORANT]
    is_edible = False

    @validator("name")
    def name_validator(cls, v):
        print("Deodorant name validator called")
        return v


class ItemRequest(BaseModel):
    __root__: Union[Banana, Sofa, Deodorant] = Field(..., discriminator="type")


@app.post("/item")
async def item(item: ItemRequest):
    print(item)
    return item
```

### Description

* Without using a request model with `__root__` and discriminator parameter, all validator methods are called for each class in the union. When using `__root__` and discriminator parameter I can correctly validate only the correct instance of the object without calling other class validators, but when using the root field, after calling the endpoint with the body parameter `item` of type `ItemRequest`, to access the correctly inferred object i have to call `item.__root__`, but when I simply return the item object, fastapi can correctly return only the content of `item.__root__`. Is there a way to automatically unpack the `__root__` contents into the object item, without manually calling `item = item.__root__` in all of my code.
* Example:
* * Input:
* * * `{
    "type": "deodorant",
    "name": "premium deodorant"
}`
* * Print output:
* * * `Deodorant name validator called`
* * * `__root__=Deodorant(type=<ItemType.DEODORANT: 'deodorant'>, name='premium deodorant', is_edible=False)`
* * Json response: 
* * * `{
    "type": "deodorant",
    "name": "premium deodorant",
    "is_edible": false
}`

### Operating System

Linux

### Operating System Details

Ubuntu 20.04.4 LTS

### FastAPI Version

0.68.2

### Python Version

3.9.7

**Comment by ADR-007** [comment]
Hi! 

I have the same problem. Any updates on it?

**Comment by ADR-007** [comment]
If you want to fix the schema at least, you can do the next:
```
def fix_root_discriminator_schema(s: dict):
    s['oneOf'] = s['properties']['root']['oneOf']
    s['discriminator'] = s['properties']['root']['discriminator']
    del s['properties']
    del s['required']

class MyModelWithRoot(BaseModel):
    class Config:
        schema_extra = fix_root_discriminator_schema
```

**Comment by YuriiMotov** [comment]
Works fine in current version (0.115.13)

Test with Pydantic V2 in details, but it also works well with Pydantic V1:

<details>

```python
from enum import Enum
from typing import Literal, Union

from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel, Field, RootModel, field_validator

app = FastAPI()


class ItemType(str, Enum):
    BANANA = "banana"
    SOFA = "sofa"
    DEODORANT = "deodorant"


class Item(BaseModel):
    type: ItemType
    name: str
    is_edible: bool


class Banana(Item):
    type: Literal[ItemType.BANANA]
    is_edible: bool = True

    @field_validator("name")
    @classmethod
    def name_validator(cls, v):
        print("Banana name validator called")
        return v


class Sofa(Item):
    type: Literal[ItemType.SOFA]
    is_edible: bool = False

    @field_validator("name")
    @classmethod
    def name_validator(cls, v):
        print("Sofa name validator called")
        return v


class Deodorant(Item):
    type: Literal[ItemType.DEODORANT]
    is_edible: bool = False

    @field_validator("name")
    @classmethod
    def name_validator(cls, v):
        print("Deodorant name validator called")
        return v


class ItemRequest(RootModel):
    root: Union[Banana, Sofa, Deodorant] = Field(..., discriminator="type")


@app.post("/item")
async def item(item: ItemRequest):
    print(item)
    return item


def test_():
    client = TestClient(app)
    response = client.post(
        "/item", json={"type": "deodorant", "name": "premium deodorant"}
    )
    assert response.status_code == 200
    assert response.json() == {
        "type": "deodorant",
        "name": "premium deodorant",
        "is_edible": False,
    }


```

</details>
